File: src/core/result.ts
import type { GameState, Player } from './types';
import type { RuleConfig } from './config';
import { ORTHO, findKing, inBoard, isGoalCell, legalMoves, opponent } from './rules';

export type WinReason = 'goal' | 'capture' | 'surround' | 'no-moves';

export interface GameResult {
  winner: Player;
  reason: WinReason;
}

const PLAYERS: Player[] = ['BLACK', 'WHITE'];

export function getResult(state: GameState, config: RuleConfig): GameResult | null {
  const n = state.board.length;

  // 0. 왕이 잡혀 사라졌으면 잡은 쪽 승리 (kingCapture 규칙)
  for (const p of PLAYERS) {
    if (!findKing(state, p)) return { winner: opponent(p), reason: 'capture' };
  }

  // 1. 왕이 목적지에 도달하면 즉시 승리
  for (const p of PLAYERS) {
    const king = findKing(state, p);
    if (king && isGoalCell(p, king, config)) return { winner: p, reason: 'goal' };
  }

  // 2. 왕 포위 패배: 상하좌우가 모두 (보드 밖 또는 상대 말)이면 그 왕의 주인이 패배
  if (config.kingSurroundLoss) {
    for (const p of PLAYERS) {
      const king = findKing(state, p);
      if (!king) continue;
      const surrounded = ORTHO.every(([dr, dc]) => {
        const r = king.r + dr;
        const c = king.c + dc;
        if (!inBoard(n, r, c)) return true;
        const piece = state.board[r][c];
        return piece !== null && piece.player !== p;
      });
      if (surrounded) return { winner: opponent(p), reason: 'surround' };
    }
  }

  // 3. 둘 수 있는 합법 수가 없으면 패배
  if (legalMoves(state, config).length === 0) {
    return { winner: opponent(state.turn), reason: 'no-moves' };
  }

  return null;
}

File: apps/ios/Sources/MongjinCore/Result.swift
import Foundation

public enum WinReason: String, Codable, Sendable {
    case goal
    case capture
    case surround
    case noMoves = "no-moves"
    case forfeit
    case timeout

    public var korean: String {
        switch self {
        case .goal: return "왕이 목적지에 도달"
        case .capture: return "상대 왕을 잡음"
        case .surround: return "상대 왕을 포위"
        case .noMoves: return "상대가 둘 수 없음"
        case .forfeit: return "상대가 항복함"
        case .timeout: return "상대가 시간 초과"
        }
    }
}

public struct GameResult: Hashable, Codable, Sendable {
    public var winner: Player
    public var reason: WinReason

    public init(winner: Player, reason: WinReason) {
        self.winner = winner
        self.reason = reason
    }

    public var label: String {
        "\(winner.korean) 승리 · \(reason.korean)"
    }
}

public func getResult(_ state: GameState, _ config: RuleConfig) -> GameResult? {
    let n = state.size

    for player in Player.allCases {
        if findKing(in: state, player: player) == nil {
            return GameResult(winner: player.opponent, reason: .capture)
        }
    }

    for player in Player.allCases {
        if let king = findKing(in: state, player: player),
           isGoalCell(player: player, coord: king, config: config) {
            return GameResult(winner: player, reason: .goal)
        }
    }

    if config.kingSurroundLoss {
        for player in Player.allCases {
            guard let king = findKing(in: state, player: player) else { continue }
            let surrounded = ortho.allSatisfy { dr, dc in
                let r = king.r + dr
                let c = king.c + dc
                if !inBoard(n, r, c) { return true }
                if let piece = state.board[r][c] {
                    return piece.player != player
                }
                return false
            }
            if surrounded {
                return GameResult(winner: player.opponent, reason: .surround)
            }
        }
    }

    if legalMoves(state, config).isEmpty {
        return GameResult(winner: state.turn.opponent, reason: .noMoves)
    }

    return nil
}
